src/kms/DeterministicCrypto.ts

```diff
@@ -0,0 +1,227 @@
+ * If `reencrypt` is true, return a failure if the field was already encrypted to the current tenant secret.
+ */
+export const decryptField = (
+    encryptedField: DeterministicEncryptedField,
+    derivedKeys: DerivedKey[] | undefined
+): Future<TenantSecurityException, DeterministicPlaintextField> => {
+    const decryptedFuture = decomposeField(encryptedField.encryptedField).flatMap((parts) => {
+        const key = derivedKeys?.find((id) => id.tenantSecretId === parts.tenantSecretId);
+        if (key === undefined) {
+            return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_DECRYPT_FAILED, "Failed deterministic decryption."));
+        } else {
+            return decryptBytes(parts.encryptedBytes, Buffer.from(key.derivedKey, "base64"));
+        }
+    });
+    return decryptedFuture.map((decrypted) => ({
+        plaintextField: decrypted,
+        derivationPath: encryptedField.derivationPath,
+        secretPath: encryptedField.secretPath,
+    }));
+};
+
+/**
+ * Deconstruct the provided encrypted field into its component parts. Separates the tenant secret ID, padding, and encrypted bytes.
+ */
+export const decomposeField = (encryptedBytesWithHeader: Buffer): Future<TenantSecurityException, DeterministicEncryptedFieldParts> => {
+    const tenantSecretIdBytes = encryptedBytesWithHeader.slice(0, 4);
+    const padding = encryptedBytesWithHeader.slice(4, 6);
+    const encryptedBytes = encryptedBytesWithHeader.slice(6);
+    if (Buffer.compare(padding, DETERMINISTIC_HEADER_PADDING) != 0) {
+        return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_HEADER_ERROR, "Failed to parse field header."));
+    }
+    const tenantSecretId = tenantSecretIdBytes.readUInt32BE();
+    return Future.of({encryptedBytes, tenantSecretId});
+};
+
+/**
+ * Attempt to AES-SIV decrypt the provided bytes using the provided key.
+ * associatedData is not used by our deterministic encryption, but is used in our unit tests of miscreant encryption.
+ */
+export const decryptBytes = (encryptedBytes: Buffer, key: Buffer, associatedData: Uint8Array[] = []): Future<TenantSecurityException, Buffer> => {
+    return Future.tryP(() => miscreant.SIV.importKey(key, "AES-SIV", cryptoProvider))
+        .flatMap((siv) => Future.tryP(() => siv.open(encryptedBytes, associatedData)).map((decrypted) => Buffer.from(decrypted)))
+        .errorMap((e) => new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_DECRYPT_FAILED, e.message));
+};
+
+/**
+ * Decrypt the provided deterministically encrypted field and re-encrypt it with the current tenant secret.
+ */
+export const rotateField = (
+    encryptedDoc: DeterministicEncryptedField,
+    derivedKeys: DerivedKey[] | undefined
+): Future<TenantSecurityException, DeterministicEncryptedField> =>
+    checkRotationFieldNoOp(encryptedDoc, derivedKeys).flatMap((noOp) => {
+        if (noOp) {
```

Comment:
Would you mind commenting here (or on the checkRotationFieldNoOp function) why we decided to do this?

Reply:
Sure, I added it on the function